### PR TITLE
feat: adding indeterminate state checkbox

### LINKED ISSUE #1720: CheckboxField - Indeterminate State
### On which framework/platform would you like to see this feature implemented?

React

### Which UI component is this feature-request for?

Other

### Please describe your feature-request in detail.

I'd like to be able to force the checkbox into an indeterminate state. See [here](https://css-tricks.com/indeterminate-checkboxes/) for example. A use-case for this is nested checkboxes where a parent's children checkboxes are not all checked.

My current workaround is to use icons, but would like to be able to use this with the CheckboxField.

### Please describe a solution you'd like.

See above.

### We love contributors! Is this something you'd be interested in working on?

- [ ] 👋 I may be able to implement this feature request.
- [ ] ⚠️ This feature might incur a breaking change.

### PR DESCRIPTION
**Pull request by zchenwei**
#### Description of changes

Add indeterminate state support to checkbox. By setting `isIndeterminate` prop to `true`, users can create a `CheckboxFiled` in indeterminate state.

Kapture 2022-08-02 at 15 18 07

***Example***
```jsx
import * as React from 'react';

import { CheckboxField, Flex, View } from '@aws-amplify/ui-react';

export const CheckboxFieldIndeterminateExample = () => {
  const [checkedItems, setCheckedItems] = React.useState([false, false]);
  const checkedItemsRef = React.useRef(null);
  const allChecked = checkedItems.every(Boolean);
  const isIndeterminate = checkedItems.some(Boolean) && !allChecked;

  if (isIndeterminate) {
    checkedItemsRef.current = [...checkedItems];
  }
  const handleAllPetsChange = () => {
    if (isIndeterminate) {
      setCheckedItems([true, true]);
    } else if (allChecked) {
      setCheckedItems([false, false]);
    } else if (checkedItemsRef.current) {
      setCheckedItems(checkedItemsRef.current);
    } else {
      setCheckedItems([true, true]);
    }
  };

  const handleCatChange = (e) => {
    const newCheckedItems = [e.target.checked, checkedItems[1]];
    if (!newCheckedItems.some(Boolean) || newCheckedItems.every(Boolean)) {
      checkedItemsRef.current = null;
    }
    setCheckedItems(newCheckedItems);
  };

  const handleDogChange = (e) => {
    const newCheckedItems = [checkedItems[0], e.target.checked];
    if (!newCheckedItems.some(Boolean) || newCheckedItems.every(Boolean)) {
      checkedItemsRef.current = null;
    }
    setCheckedItems(newCheckedItems);
  };

  return (
    <Flex direction="column" gap="0">
      <CheckboxField
        name="all-pets"
        label="All Pets"
        value="allPets"
        checked={allChecked}
        isIndeterminate={isIndeterminate}
        onChange={handleAllPetsChange}
      />
      <View paddingLeft="25px">
        <CheckboxField
          name="cat"
          label="Cat"
          value="cat"
          checked={checkedItems[0]}
          onChange={handleCatChange}
        />
        <CheckboxField
          name="dog"
          label="Dog"
          value="dog"
          checked={checkedItems[1]}
          onChange={handleDogChange}
        />
      </View>
    </Flex>
  );
};
```

#### Issue #, if available

Indeterminate checkbox support. Fixes #1720 

#### Description of how you validated changes

Add new unit test to cover the new feature

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are updated
- [x] No side effects or [`sideEffects`](https://github.com/aws-amplify/amplify-ui/blob/main/packages/react/CONTRIBUTING.md#code-standards) field updated
- [x] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/react/src/primitives/Checkbox/Checkbox.tsx [hunk 1]
```diff
@@ -21,11 +22,11 @@ const CheckboxPrimitive: Primitive<CheckboxProps, 'input'> = (
     defaultChecked,
     hasError,
     isDisabled,
+    isIndeterminate,
```
Comment (calebpollman) [inline packages/react/src/primitives/Checkbox/Checkbox.tsx:25]: Will this conflict with any existing props? For example `defaultChecked`

**Review by calebpollman — COMMENTED** [review]
Looks like a good start to me 😄  Left some questions, let me know your thoughts

File: packages/react/src/primitives/Checkbox/Checkbox.tsx [hunk 2]
```diff
@@ -21,11 +22,11 @@ const CheckboxPrimitive: Primitive<CheckboxProps, 'input'> = (
...
     label,
     labelHidden,
     labelPosition,
     onChange: onChangeProp,
-    size,
```
[the remark this answers is not in the document]
Comment (zchenwei) [reply packages/react/src/primitives/Checkbox/Checkbox.tsx:28]: Just cleaning up the useless code. We have remove `size` prop on icon long time a ago and let it scale in proportion to its font size.

File: packages/react/src/primitives/Checkbox/Checkbox.tsx (hunk 1 above)
Comment (zchenwei) [reply packages/react/src/primitives/Checkbox/Checkbox.tsx:25]: It complies with native HTML spec. `indeterminate` state will always have the precedence over other states.

File: packages/react/src/primitives/Checkbox/Checkbox.tsx [hunk 3]
```diff
@@ -82,6 +91,27 @@ const CheckboxPrimitive: Primitive<CheckboxProps, 'input'> = (
     )
   );
 
+  const renderedIcon = isIndeterminate ? (
```
[the remark this answers is not in the document]
Comment (zchenwei) [reply packages/react/src/primitives/Checkbox/Checkbox.tsx:94]: Sounds good

File: packages/react/src/primitives/Checkbox/Checkbox.tsx [hunk 4]
```diff
@@ -104,6 +134,7 @@ const CheckboxPrimitive: Primitive<CheckboxProps, 'input'> = (
         <Input
           checked={checked}
           className={ComponentClassNames.CheckboxInput}
+          data-id={dataId}
```
[the remark this answers is not in the document]
Comment (zchenwei) [reply packages/react/src/primitives/Checkbox/Checkbox.tsx:137]: used target this element with `document.querySelector` on line 53

**Review by calebpollman — DISMISSED** [review]
LGTM 🚢

File: .changeset/gorgeous-rocks-carry.md [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+export const CheckboxFieldIndeterminateExample = () => {
+  const [checkedItems, setCheckedItems] = React.useState([false, false]);
+  const allChecked = checkedItems.every(Boolean);
+  const isIndeterminate = checkedItems.some(Boolean) && !allChecked;
+
+  return (
+    <Flex direction="column" gap="0">
+      <CheckboxField
+        name="parent-checkbox"
```
Comment (hbuchel) [inline .changeset/gorgeous-rocks-carry.md:24]: Nit: Can we use realistic labels/names? i.e. 
```
All animals
- Dog
- Cat
```

**Review by ioanabrooks — APPROVED** [review]

**Review by jacoblogan — APPROVED** [review]

**Review by hbuchel — APPROVED** [review]

**Pull request merged**